lgc/patch/PatchInitializeWorkgroupMemory.cpp

```diff
@@ -0,0 +1,176 @@
+// Executes this LLVM patching pass on the specified LLVM module.
+//
+// @param [in/out] module : LLVM module to be run on
+bool PatchInitializeWorkgroupMemory::runOnModule(Module &module) {
+  LLVM_DEBUG(dbgs() << "Run the pass Patch-Initialize-Workgroup-Memory\n");
+
+  PipelineState *pipelineState = getAnalysis<PipelineStateWrapper>().getPipelineState(&module);
+  // This pass works on compute shader.
+  if (!pipelineState->hasShaderStage(ShaderStageCompute))
+    return false;
+
+  SmallVector<GlobalVariable *> zeroGlobals;
+  for (GlobalVariable &global : module.globals()) {
+    if (global.hasInitializer() && global.getInitializer()->isNullValue())
+      zeroGlobals.push_back(&global);
+  }
+
+  if (zeroGlobals.empty())
+    return false;
+
+  Patch::init(&module);
+  m_shaderStage = ShaderStageCompute;
+  m_entryPoint = (&getAnalysis<PipelineShaders>())->getEntryPoint(static_cast<ShaderStage>(m_shaderStage));
+  BuilderBase builder(*m_context);
+  Instruction *insertPos = &*m_entryPoint->front().getFirstInsertionPt();
+  builder.SetInsertPoint(insertPos);
+
+  for (auto global : zeroGlobals) {
+    Type *globalTy = global->getType();
+    // Create new global with null initializer to replace the original global. The global with zeroinitializer is not
+    // supported by back-end
+    GlobalVariable *newGlobal =
+        new GlobalVariable(module, globalTy->getPointerElementType(), false, GlobalValue::ExternalLinkage, nullptr,
+                           "lds", nullptr, GlobalValue::NotThreadLocal, globalTy->getPointerAddressSpace());
+    newGlobal->setAlignment(MaybeAlign(global->getAlignment()));
+    global->replaceAllUsesWith(newGlobal);
+
+    Type *ptrElemTy = globalTy->getPointerElementType();
+    SmallVector<Value *> indices;
+    if (ptrElemTy->isAggregateType())
+      indices.push_back(builder.getInt32(0));
+    initializeWithZero(newGlobal, ptrElemTy, indices, builder);
+
+    global->eraseFromParent();
+  }
+
+  return true;
+}
+
+// =====================================================================================================================
+// Recursively search the single value for a given type and write the corresponding zero value to the specified address
+// on LDS.
+
+//
+// @param pointer : The pointer to the address on LDS
+// @param valueTy : The checking value type
+// @param indices : The index list required for an aggregative type
+// @param builder : BuilderBase to use for instruction constructing
+void PatchInitializeWorkgroupMemory::initializeWithZero(Value *pointer, Type *valueTy,
+                                                        SmallVectorImpl<Value *> &indices, BuilderBase &builder) {
+  if (valueTy->isSingleValueType()) {
+    Value *zero = Constant::getNullValue(valueTy);
+    unsigned alignment = valueTy->getPrimitiveSizeInBits() / 8;
+    if (!isPowerOf2_32(alignment))
+      alignment = (alignment & (alignment - 1)) * 2;
```

Comment:
I think this could be changed to alignment = alignTo(alignment, 2)

Reply:
`alignTo(alignment, 2)` returns a multiple of  2. But I want to make the alignment be power of 2.

Reply:
You can choose: NextPowerOf2, PowerOf2Floor, or PowerOf2Ceil. They can achieve what you need.

Reply:
Thanks! Done.